### PR TITLE
Add Operation: `Same`

### PR DESCRIPTION
**Pull request by AlexandruGG**
This PR:

* [x] Provides a new operation `Same` which allows determining if two collections are the same
* [x] Allows custom item comparison via a callable -> if this is not provided, by default we consider two collections the same if the have the same items and keys, in the same order
* [x] Has unit tests (phpspec)
* [x] Has static analysis tests (psalm, phpstan)
* [x] Has documentation

Follows #152.
Related to https://github.com/loophp/collection/discussions/135#discussioncomment-1016598.

### CODE REVIEW

**Comment by AlexandruGG** [comment]
@drupol hey, this PR still needs some work but let me know if you're happy with the implementation itself. Very similar to Equals, with a few minor differences. It's actually simpler.

**Comment by AlexandruGG** [comment]
Also, should we just ignore this phpcs rule for the whole project, or at least `src/Operation`? 😄

**Comment by drupol** [comment]
About PHPCS: Let's add the rule in the file header for now.

Regarding the algorithm used in here, I think we could get rid of local variables using early returns.

**Comment by AlexandruGG** [comment]
> About PHPCS: Let's add the rule in the file header for now.

Yep that's what I will do for this PR, but if we know for sure we will have this issue in many Operations due to currying and type annotations, would it not be good to disable it at least for that namespace? I think it makes the code look a bit more "messy" with the ignores in each file

**Comment by AlexandruGG** [comment]
> Regarding the algorithm used in here, I think we could get rid of local variables using early returns.

Yep that's what I had initially but ran into some php-cs-fixer issues, have a look at my comments here: https://gist.github.com/AlexandruGG/6b201e47bc4e332f72d7080132cc944a

**Comment by AlexandruGG** [comment]
> > Regarding the algorithm used in here, I think we could get rid of local variables using early returns.
> 
> Yep that's what I had initially but ran into some php-cs-fixer issues, have a look at my comments here: https://gist.github.com/AlexandruGG/6b201e47bc4e332f72d7080132cc944a

I updated the PR with an alternative. It's a bit more verbose but it works if we want to avoid the local variable

**Comment by drupol** [comment]
@AlexandruGG I pushed a minor commit so we don't forget to fix later on!

**Review by drupol — DISMISSED** [review]
I will be busy today and I don't know if I will be available that much, but the PR is fine for me, feel free to merge it when the tests are done and you feel like it's good.

**Comment by AlexandruGG** [comment]
> @AlexandruGG I pushed a minor commit so we don't forget to fix later on!

@drupol thank you! Weirdly enough, I couldn't manage to reproduce that issue in a separate project, no matter how hard I tried 😄

**Comment by AlexandruGG** [comment]
> I will be busy today and I don't know if I will be available that much, but the PR is fine for me, feel free to merge it when the tests are done and you feel like it's good.

@drupol thank you! Sorry but the last commit dismissed your review and GitHub won't let me merge it without the approval 😄

**Review by drupol — APPROVED** [review]
Quickly passing to approve, thanks for your work :) if you have time to check #153 as well it could be nice.

**Comment by AlexandruGG** [comment]
> Quickly passing to approve, thanks for your work :) if you have time to check #153 as well it could be nice.

Thank you! Sure I'll have a look at that one this evening 😁

**Pull request merged**
